### PR TITLE
Add memory staging to stencil chain

### PR DESCRIPTION
**Pull request by abadams**
Makes the manual GPU schedule about 2x faster by replacing loads from global with vector loads from shared.

### CODE REVIEW

**Comment by abadams** [comment]
Actually half of the gain (1.5ms) was from #5177

**Comment by abadams** [comment]
review ping

File: apps/stencil_chain/stencil_chain_generator.cpp [hunk 1]
```diff
@@ -50,23 +50,37 @@ class StencilChain : public Halide::Generator<StencilChain> {
...
                     .tile(xi, yi, xii, yii, 2, 2)
                     .unroll(xii)
                     .unroll(yii);
+                prev.in()
+                    .compute_at(s, x)
+                    .tile(prev.args()[0], prev.args()[1], xi, yi, 2, 2)
+                    .vectorize(xi)
+                    .unroll(yi)
+                    .gpu_threads(prev.args()[0], prev.args()[1]);
+                prev.in()
```
Comment (steven-johnson) [inline apps/stencil_chain/stencil_chain_generator.cpp:76]: ...I have no idea what `.in().in()` implies for scheduling. 

(In general, no pun intended, I have never fully understood the practical use cases for `.in()` and would love to see expanded tutorials on it.)

Comment (abadams) [reply apps/stencil_chain/stencil_chain_generator.cpp:76]: This seems like a good place for some explanatory comments then...

Comment (abadams) [reply apps/stencil_chain/stencil_chain_generator.cpp:95]: PTAL and tell me if the comments clear things up for you

Comment (steven-johnson) [reply apps/stencil_chain/stencil_chain_generator.cpp:95]: Yes, it helps immensely. I had no idea that's how `.in()` could be used. It sounds tremendously useful for situations like this. I'm going to open an issue for "add tutorial lesson for .in()"

**Pull request merged**